Trigger GC for actors when they tell the cycle detector they're blocked

dipinhora opened the pull request:
Prior to this commit, if an actor blocked, it did not run GC to free
any memory it no longer needed. This would result in blocked actors
holding on to (potentially lots of) memory unnecessarily.

This commit causes GC to be triggered when the cycle detector asks
an actor if it is blocked and the actor responds telling the cycle
detector that it is blocked. This should result in memory being
held by blocked actors to be freed more quickly even if the cycle
detector doesn't end up detecting a cycle and reaping the actors.

This will force a GC for an actor based on the following three things:

* The actor processed at least one message since it's last GC (i.e. it did some work [GC acquire/release message or app message)
* The actor's heap is greater than 0 (i.e. it has memory that could potentially be freed)
* The actor is blocked and is about to tell the cycle detector that it is blocked (i.e. it thinks it has no more work to do at the moment)

The sequence of events for GC'ing when sending a block message to the CD is:

1. actor gets a message from another actor
1. gets rescheduled because it processed an application message
1. next run has an empty queue (and the actor gets marked internally as blocked but doesn't send a block message to the CD)
1. some time passes and the CD eventually asks the actor if it is blocked
1. the actor garbage collects because of this change before sending the block message to the CD (to prevent race conditions)
1. the actor responds to the CD by sending a block message

This shouldn't be a performance hit because step 4 is based on how often the CD runs (not very often) along with the fact that the CD doesn't ask all actors it knows about if they're blocked on every run, it asks them in batches instead and so step 4 will not occur very frequently for any actor even if steps 1 - 3 happen regularly.

SeanTAllen commented:
@dipinhora looks like there is a race condition here that is triggering an assert condition that could result in a segfault. otherwise, the general idea looks good.

dipinhora commented:
@SeanTAllen is there a reason you believe the race condition is a result of the changes in this PR? i see the CI failure with the assertion on shutdown as part of the `alpine-lib-llvm-debug-static` circle ci job but nothing in this PR is touching any of that code or impacting it meaningfully in any way.

SeanTAllen commented:
Is that not a new assertion failure @dipinhora? I don't believe I've seen it before.

dipinhora commented:
@SeanTAllen it is the first time i've seen it also. however, that doesn't mean it is caused by the changes in this PR.

SeanTAllen commented:
I believe that view->perceived is related to the cycle detector and gc which suggests to me that it is related. Perhaps we should run the debug tests repeatedly and see if we can get it to trigger with this change and without?

dipinhora commented:
`view->perceived` has to do with the internal workings of the cycle detector. the cycle detector has nothing to do with actor gc. this pr is only tangentially related to the cycle detector. it does not modify any logic within the cycle detector. nor does it introduce a new message type for the cycle detector and actors to deal with. it causes gc to be triggered for an actor when a specific message type is received by an actor (which happens to come from the cycle detector). while this pr may have exposed the race condition that caused the assertion to fail, it is not the cause of it.

note: i am not suggesting that the issue exposed by the failed circle ci job not be investigated and fixed. i believe it should be investigated and fixed; just not as part of this pr.

sylvanc commented:
If this change were to happen, the GC needs to happen on the actor's current turn. This implementation sets the actor up to GC on it's _next_ turn, i.e. after it has unblocked.

The heuristic should not be based on messages received since last GC, but rather a more aggressive form of the "normal" GC heuristic. That is, "collect at 25% growth instead of 50% growth" (just an example).

SeanTAllen commented:
@dipinhora shall I pick up work on this PR?

dipinhora commented:
@SeanTAllen rebased/updated with comments to explain what/why. Let me know if you have any questions about the changes or the comments.

SeanTAllen commented:
@dipinhora do you think this should have a changelog entry and release notes?

dipinhora commented:
@SeanTAllen not sure.. this isn't really a user facing thing.. but it is a runtime behavior change similar in some ways to your recent short lived actor reaping changes.. and i think that has a changelog/release notes so i suppose this should also? 8*/

SeanTAllen commented:
Not user facing but..

Perhaps of interest.

I'm leaning towards no

dipinhora commented:
works for me

SeanTAllen commented:
Current CI failures in this PR are related to a bug that is fixed in https://github.com/ponylang/ponyc/pull/3666.

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -90,6 +90,22 @@ static void send_unblock(pony_actor_t* actor)
+  pony_assert(ctx->current == actor);
+  ponyint_cycle_block(actor, &actor->gc);
+
+  // trigger a GC if we're sending a block message to the CD
+  // GC will get run next time `try_gc` is called which should
+  // happen once all messages get processed in the messageq
+  // at the time `pony_actor_run` was called if not earlier
+  // when an app message is processed
```
```suggestion
  // Trigger garbage collection. GC will get run next time `try_gc` is called 
  // which should happen once all messages get processed in the messageq
  // at the time `pony_actor_run` was called if not earlier
  // when an app message is processed
```

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -866,7 +879,10 @@ void* pony_alloc_large_final(pony_ctx_t* ctx, size_t size)
-  ctx->current->heap.next_gc = 0;
+
+  // only trigger gc if actor allocated something on the heap
+  if(ctx->current->heap.used > 0)
+    ctx->current->heap.next_gc = 0;
```
This isn't technically needed as the logic that decides if we gc or not in `ponyint_heap_startgc` is:

```c
  if(heap->used <= heap->next_gc)
    return false;
```

so setting next_gc to 0 when heap.used is 0 will not result in gc happening.

I thin this should be removed as it is extra cycles being used.

SeanTAllen replied:
Is the logic that you really wanted here "only gc if we have allocated anything since the last time"? I think that is what you want.

SeanTAllen replied:
actually i see what you are doing here. if we did any memory usage at all, even 1 byte, then try gc'ing.

but, isn't 0 still something that could be of issue? what if we are holding a ton of stuff that could be freed but didn't allocate anything else on the heap. is that possible?

dipinhora replied:
Your first comment was correct. I forgot about the same check already existed in `ponyint_heap_startgc`.

SeanTAllen reviewed (changes requested):
There's a couple small changes I would like to see to this as I noted.

Otherwise, now that I fully understand the changes that you made to the cycle detector last year Dipin, I think this is a reasonable PR.

I can see some scenarios where this might result in better memory usage.

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -418,9 +430,12 @@ bool ponyint_actor_run(pony_ctx_t* ctx, pony_actor_t* actor, bool polling)
-      ponyint_cycle_block(actor, &actor->gc);
+      send_block(ctx, actor);
+
+      // Try and run GC because we're blocked and sending a block message
+      // to the CD and we're past the normal point at which `try_gc` is
+      // called as part of this function
```
Can you change this message to more of "the why" rather than the "what".

Something like "we will try to free any memory we can as we are blocked and might hold on to that memory for a long time."

dipinhora replied:
updated

SeanTAllen commented:
Reasons I think this is probably a good trade off between performance and memory usage.

The gc is only done IF the actor in question didn't receive an application message. So the only time this should happen is when an actor receives "is blocked" message from cycle detector and there were no messages from other actors during the run.

This should I suspect be rare. It is likely in that case that the actor might hold on to any garbage it has for a long period of time. That could be a large amount of memory.

This will be a performance hit in a scenario like:

- actor gets a message from another actor
- gets rescheduled because it processed an application message
- next run has an empty queue
- garbage collects because of this change

SeanTAllen commented:
@dipinhora this raises an interesting gc edge-case. for me at least, is the following true currently (i think it is).

- an actor allocates say 1 gig of memory. it is using it all, gc is run, it still holds 1 gig.
- actor receives a message telling it that it isn't needed anymore. 
- actor doesn't allocate enough memory to push heap_used past next_gc
- actor holds on to unneeded memory for potentially a very long time

with your change this problem still exists when the cycle detector is off. and with your change, can this problem still exist because heap used was 0? 

something like we have a field that is (TonsOfMemory | None) and we set to None, so heap used is still 0 and we don't try to garbage collect "TonsOfMemory".

dipinhora commented on src/libponyrt/actor/actor.c:
```diff
@@ -90,6 +90,22 @@ static void send_unblock(pony_actor_t* actor)
+  // We're blocked, send block message.
+  set_flag(actor, FLAG_BLOCKED_SENT);
+  set_flag(actor, FLAG_CD_CONTACTED);
+  pony_assert(ctx->current == actor);
```
I moved the assertion up a few lines. i kept it because the original code path in `handle_message` had it and it's only asserted in a debug build so doesn't impact performance in release builds. The reason not use use `ctx->current` when the same value is in `actor` is because `actor` should already be available as a value in a register while loading from `ctx->current` could result in requiring a memory load. Note: the former may be incorrect but that is my limited understanding of how compilers tend to translate C code to assembly.

dipinhora commented:
> Reasons I think this is probably a good trade off between performance and memory usage.
> 
> The gc is only done IF the actor in question didn't receive an application message. So the only time this should happen is when an actor receives "is blocked" message from cycle detector and there were no messages from other actors during the run.
> 
> This should I suspect be rare. It is likely in that case that the actor might hold on to any garbage it has for a long period of time. That could be a large amount of memory.
> 
> This will be a performance hit in a scenario like:
> 
> * actor gets a message from another actor
> * gets rescheduled because it processed an application message
> * next run has an empty queue
> * garbage collects because of this change

I don't believe this will be a performance hit. The performance hit scenario you've outline is missing an important detail. The following is what would happen in your scenario with the detail added:

The scenario would be:

1. actor gets a message from another actor
1. gets rescheduled because it processed an application message
1. next run has an empty queue (and the actor gets marked internally as blocked but doesn't send a block message to the CD)
1. some time passes and the CD eventually asks the actor if it is blocked
1. the actor responds to the CD by sending a block message
1. the actor garbage collects because of this change after sending the block message to the CD

I don't believe this will be a performance hit because step 4 is based on how often the CD runs (not very often) along with the fact that the CD doesn't ask all actors it knows about if they're blocked on every run, it asks them in batches instead and so step 4 will not occur very frequently for any actor even if steps 1 - 3 happen regularly.

SeanTAllen commented:
can you rebase against master to pick up the race condition fix?

dipinhora commented:
rebased

SeanTAllen commented:
@dipinhora I understand the scenario you outlined. I think there is a performance hit there "extra gc time", but I think it's small in almost all cases and worth the memory reclaimation.

dipinhora commented:
> @dipinhora this raises an interesting gc edge-case. for me at least, is the following true currently (i think it is).
> 
> * an actor allocates say 1 gig of memory. it is using it all, gc is run, it still holds 1 gig.
> * actor receives a message telling it that it isn't needed anymore.
> * actor doesn't allocate enough memory to push heap_used past next_gc
> * actor holds on to unneeded memory for potentially a very long time
> 
> with your change this problem still exists when the cycle detector is off. and with your change, can this problem still exist because heap used was 0?
> 
> something like we have a field that is (TonsOfMemory | None) and we set to None, so heap used is still 0 and we don't try to garbage collect "TonsOfMemory".

Yes, this will still exist when the CD is off as this change only addresses for when the CD is on.

I don't believe the edge case you've described is a concern because `heap.used` only gets set to `0` when GC is run and the only way it can be `0` is because the GC freed all memory (or there's a bug). All other normal actor activity (including something like `we have a field that is (TonsOfMemory | None) and we set to None`) can only result in `heap.used` increasing (or remaining the same).

SeanTAllen commented:
@dipinhora can you rebase this again when you get a chance? once that is done and it passes CI, I'll merge this.

dipinhora commented:
@SeanTAllen rebased

SeanTAllen commented:
I totally missed that this was rebased ages ago, is this ready @dipinhora?

dipinhora commented:
> I totally missed that this was rebased ages ago, is this ready @dipinhora?

In theory (assuming the `freebsd CI failure` is a spurious/known thing.

> If yes, can you update the first comment to capture some of the additional information that came as conversation on this PR and I'll use that when squashing this?

Yeah, i'll try and update that first comment to capture more info. Will drop a note when that's done.

dipinhora commented:
`freebsd` failure:

```
promise-flatten-next-regression-3856 (5450 ms): testing failed: signal 11
```

i don't believe it's related to the changes in this PR and i'm going to retry.

SeanTAllen commented:
Isn't signal 11 a segfault?

dipinhora commented:
yep, sig 11 is a segfault..

SeanTAllen commented:
That's not good!

dipinhora commented:
i agree.. segfaults aren't good..

dipinhora commented:
Retrying fixed the `freebsd` CI failure..

so, the question is, was the `segfault` a result of the changes in this PR or not? any thoughts?

SeanTAllen commented:
I've never seen that error before but it could be an existing issue we happened to see. I'm ok with merging and keeping an eye out for similar problems.

dipinhora commented:
hmm... if it's something you haven't seen before then lets wait on the merge while i pretend to be a computer and try and figure out if i've accidentally introduced a race condition.. and if i still can't think of anything then maybe we go with the merge and monitor approach..

SeanTAllen commented:
Sounds like a plan

dipinhora commented:
In theory moving the GC to occur before sending the `block` message should prevent race conditions that would exist if GC is done after sending a `block` message.

This is almost ready (i still need to update that first comment)... i'll drop a note whenever that's done.

dipinhora commented:
@SeanTAllen i've updated the first comment.. hopefully it's a good enough summary of the conversation..

jemc commented:
@SeanTAllen - does your approval from 2020 still stand or do you want to review again?

dipinhora commented:
rebased to resolve conflict

dipinhora commented:
anyone have any ideas why the windows CI build would time out running tests? i don't believe it's related to this change..

SeanTAllen commented:
The timeouts aren't happening with other PRs so I think it is related.

dipinhora commented:
Considering that there were no timeouts prior to the rebase (all tests passed), and that this PR changes absolutely nothing related to atomics used by the schedulers or the quiescence logic, and only contains a minor change related to triggering an extra GC cycle in certain scenarios, i'm finding it hard to understand how this PR is related to the timeouts.. i'm happy to look deeper if you can point to specific changes in the PR that you think might be suspect but i'm drawing a blank otherwise..

dipinhora commented:
another datapoint: the windows CI timed out in the last commit for PR https://github.com/ponylang/ponyc/pull/4164 also.

SeanTAllen commented:
I'm most interested now in why this one PR always exhibits the windows hang. I suspect that whatever the occasional trigger elsewhere, it always happens with these changes.

dipinhora commented:
> I'm most interested now in why this one PR always exhibits the windows hang. I suspect that whatever the occasional trigger elsewhere, it always happens with these changes.

Not sure why you think `this one PR always exhibits the windows hang` when, as i stated in my comment yesterday, `there were no timeouts prior to the rebase (all tests passed)` (maybe you're getting this PR confused with https://github.com/ponylang/ponyc/pull/4151 where we saw many windows hangs?).. Additionally, the issue that this PR exhibited previously was a `FreeBSD` crash (nothing Windows related) which was resolved..

Regardless, i've rebased again to pick up latest `main` changes.. Let's roll the dice and see if the Windows hang occurs again or not.

SeanTAllen commented:
Because this now fails every time I restart it. It's not occasional.

dipinhora commented:
> every time I restart it

That sounds like something important that i wasn't aware of (or somehow missed) previously..

dipinhora commented:
looks like the windows Ci timeout issue isn't limited to this PR or https://github.com/ponylang/ponyc/pull/4164.. it just happened on https://github.com/ponylang/ponyc/pull/3282 also.

dipinhora commented:
Funny.. it passed successfully this time on this PR..
